metaflow/plugins/secrets/secrets_decorator.py

```diff
@@ -0,0 +1,197 @@
+            secrets_backend_type = get_default_secrets_backend_type()
+        else:
+            secrets_backend_type = secret_spec_dict["type"]
+            if type(secrets_backend_type) != str:
+                raise MetaflowException(
+                    "Bad @secrets specification - 'type' must be a string - found %s"
+                    % type(secrets_backend_type)
+                )
+        secret_id = secret_spec_dict.get("id")
+        if type(secret_id) != str:
+            raise MetaflowException(
+                "Bad @secrets specification - 'id' must be a string - found %s"
+                % type(secret_id)
+            )
+        options = secret_spec_dict.get("options", {})
+        if not isinstance(options, dict):
+            raise MetaflowException(
+                "Bad @secrets specification - 'option' must be a dict - found %s"
+                % type(options)
+            )
+        return SecretSpec(secrets_backend_type, secret_id=secret_id, options=options)
+
+
+def validate_env_vars_across_secrets(all_secrets_env_vars):
+    vars_injected_by = defaultdict(str)
+    for secret_spec, env_vars in all_secrets_env_vars:
+        for k, v in env_vars.items():
+            if k in vars_injected_by:
+                raise MetaflowException(
+                    "Secret '%s' will inject '%s' as env var, but it also by '%s'"
+                    % (secret_spec.secret_id, k, vars_injected_by[k])
+                )
+            vars_injected_by[k] = secret_spec.secret_id
+
+
+def validate_env_vars_vs_existing_env(all_secrets_env_vars):
+    for secret_spec, env_vars in all_secrets_env_vars:
+        for k, v in env_vars.items():
+            if k in os.environ:
+                raise MetaflowException(
+                    "Secret %s will inject %s as env var, but it already exists in env"
+                    % (secret_spec.secret_id, k)
+                )
+
+
+def validate_env_vars(env_vars):
+    for k, v in env_vars.items():
+        if type(k) != str:
+            raise ValueError("Found non-str key %s (%s)" % (str(k), type(k)))
+        if type(v) != str:
+            raise ValueError("Found non-str value %s (%s)" % (str(v), type(v)))
+        if not re.fullmatch("[a-zA-Z_][a-zA-Z0-9_]*", k):
+            raise ValueError("Found invalid env var name '%s'." % k)
+
+
+def get_secrets_backend_provider(secrets_backend_type):
+    try:
+        return SECRETS_BACKEND_PROVIDERS[secrets_backend_type]
+    except KeyError:
+        raise MetaflowException(
+            "Unknown secrets backend type %s" % secrets_backend_type
+        )
+
+
+class SecretsDecorator(StepDecorator):
+    """
+    Specifies secrets to be retrieved and injected as environment variables prior to the execution of a step.
+
+    Parameters
+    ----------
+    vars : List[Union[str,dict]]
+        List of secret specs, defining how the secrets are to be retrieved
+    """
+
+    name = "secrets"
+    defaults = {"vars": []}
+
+    def runtime_step_cli(
```

Comment:
shouldn't this be `task_pre_step` so that secret access happens within the task execution context. Currently secrets are accessed before the task is launched.